File: 76-minimum-window-substring/minimum-window-substring.py
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        if len(s)<len(t):
            return ""
        tchar={}
        for char in t:
            if char not in tchar:
                tchar[char]=1
            else:
                tchar[char]+=1
        l,r=0,0
        count=0
        ans=float('inf')
        res=""
        window={}
        while r<len(s):
            window[s[r]]=window.get(s[r],0)+1

            if s[r] in tchar and tchar[s[r]]==window[s[r]]:
                count+=1
            while count==len(tchar):
                minlen=r-l+1
                if minlen<ans:
                    ans=minlen
                    res=s[l:r+1]
                window[s[l]]-=1
                if s[l] in tchar and window[s[l]]<tchar[s[l]]:
                    count-=1
                l+=1
            r+=1
        return res


            


        

File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int j=0;
        unordered_map<char,int> mp1,mp2;
      
        for(char c:t){
            mp1[c]++;
        }
        int n=t.size(),k=0;
        if(n>s.size()) return "";
        int len=INT_MAX;
        int count=0;
        for(int i=0;i<s.size();i++){
            mp2[s[i]]++;
            if(mp1.count(s[i]) && mp2[s[i]]<=mp1[s[i]]){
                count++;

            }
            while(count==n ){
                if(i-j+1<len){
                    len=i-j+1;
                    k=j;
                }
                mp2[s[j]]--;
                if(mp1.count(s[j]) && mp2[s[j]]<mp1[s[j]]){
                    count--;
                }
                j++;
            }

        }
        
        return len!=INT_MAX?s.substr(k,len):"";
    }
};